### PR TITLE
[v0.8] Move self test logic to crate `mbedtls-selftest`

### PR DESCRIPTION
**Pull request by Taowyoo**
# mbedtls-selftest

This Rust crate is designed for separating self-test code that needs to export Rust
functions and define C functions to be used by C `mbedtls`. By separating this code,
different versions of Rust `mbedtls` crates can be used within a single crate, which
helps to solve link name conflict errors.

**Note**: Although multiple versions of Rust `mbedtls` crates can be used within a
single crate, only one `mbedtls-selftest` and one `mbedtls-sys-auto` crate can be
used since they are built as native libraries.

## Changes need attention
- Pump version of `mbedtls` to `0.8.3`
- Cherry picked PR: #180, #206, #214 to fix CI and compile errors
- Add patch in workspace `Cargo.toml` to ensure `mbedtls-selftest` use local `mbedtls-sys-auto` crate

### CODE REVIEW

**Comment by Taowyoo** [comment]
bors try

File: mbedtls/Cargo.toml [hunk 1]
```diff
@@ -38,9 +38,13 @@ chrono = "0.4"
 [dependencies.mbedtls-sys-auto]
 version = "2.25.0"
 default-features = false
-features = ["custom_printf", "trusted_cert_callback", "threading"]
+features = ["trusted_cert_callback", "threading"]
 path = "../mbedtls-sys"
 
+[dependencies.mbedtls-selftest]
+version = "0.1"
+path = "../mbedtls-selftest"
```
Comment (jethrogb) [inline mbedtls/Cargo.toml:46]: Don't add it as a path dependency for the older branches

Comment (Taowyoo) [reply mbedtls/Cargo.toml]: Removed

**Comment by Taowyoo** [comment]
bors try

**Comment by Taowyoo** [comment]
bors try

**Comment by jethrogb** [comment]
Backport https://github.com/fortanix/rust-mbedtls/pull/243 as well?

**Comment by Taowyoo** [comment]
> Backport #243 as well?

Oh, yes, it's also needed for 0.8.X

**Comment by Taowyoo** [comment]
bors cancel

**Comment by Taowyoo** [comment]
bors try

**Comment by Taowyoo** [comment]
bors try

**Comment by Taowyoo** [comment]
bors cancel

**Comment by Taowyoo** [comment]
bors try

**Comment by Taowyoo** [comment]
Close this PR, and continue work with remain in:
- https://github.com/fortanix/rust-mbedtls/pull/259

**Pull request closed without merging**